resources/smpp/ra/src/main/java/org/restcomm/slee/resource/smpp/SmppSessionsImpl.java

```diff
@@ -77,79 +77,59 @@ public SmppTransaction sendRequestPdu(Esme esme, PduRequest request, long timeou
-					timeoutMillis, false);
-		} catch (RecoverablePduException e) {
-			this.smppServerResourceAdaptor.endActivity(smppServerTransaction);
-			throw e;
-		} catch (UnrecoverablePduException e) {
-			this.smppServerResourceAdaptor.endActivity(smppServerTransaction);
-			throw e;
-		} catch (SmppTimeoutException e) {
-			this.smppServerResourceAdaptor.endActivity(smppServerTransaction);
-			throw e;
-		} catch (SmppChannelException e) {
-			this.smppServerResourceAdaptor.endActivity(smppServerTransaction);
-			throw e;
-		} catch (InterruptedException e) {
-			this.smppServerResourceAdaptor.endActivity(smppServerTransaction);
-			throw e;
-		}
+		SmppSendingTask task = new SmppSendingTask(esme, request, timeoutMillis, null, smppServerTransaction);
+		sender.offer(task);
 
 		return smppServerTransaction;
 	}
 
 	@Override
-	public void sendResponsePdu(Esme esme, PduRequest request, PduResponse response) throws RecoverablePduException,
-			UnrecoverablePduException, SmppChannelException, InterruptedException {
+	public void sendResponsePdu(Esme esme, PduRequest request, PduResponse response) throws UnrecoverablePduException {
 
 		SmppTransactionImpl smppServerTransactionImpl = (SmppTransactionImpl) request.getReferenceObject();
 
-		try {
-			DefaultSmppSession defaultSmppSession = esme.getSmppSession();
+		DefaultSmppSession defaultSmppSession = esme.getSmppSession();
 
-			if (defaultSmppSession == null) {
-				throw new NullPointerException("Underlying SmppSession is Null!");
-			}
+		if (defaultSmppSession == null) {
+			throw new NullPointerException("Underlying SmppSession is Null!");
+		}
 
-			if (request.getSequenceNumber() != response.getSequenceNumber()) {
-				throw new UnrecoverablePduException("Sequence number of response is not same as request");
-			}
-			defaultSmppSession.sendResponsePdu(response);
-		} finally {
-			
-			SmppSessionCounters smppSessionCounters = esme.getSmppSession().getCounters();
-			SmppTransactionImpl smppTransactionImpl = (SmppTransactionImpl)request.getReferenceObject();
-			long responseTime = System.currentTimeMillis() - smppTransactionImpl.getStartTime();
-			this.countSendResponsePdu(smppSessionCounters, response, responseTime, responseTime);
-			
-			if (smppServerTransactionImpl == null) {
-				tracer.severe(String.format("SmppTransactionImpl Activity is null while trying to send PduResponse=%s",
-						response));
-			} else {
-				this.smppServerResourceAdaptor.endActivity(smppServerTransactionImpl);
-			}
+		SenderThread sender = esmeSenderThreads.get(esme.getName());
+		if (sender == null) {
+			throw new IllegalStateException("Esme sender not found");
 		}
 
+		if (request.getSequenceNumber() != response.getSequenceNumber()) {
+			throw new UnrecoverablePduException(
+					"Sequence number of response is not same as request");
+		}
+
+		SmppSendingTask task = new SmppSendingTask(esme, request, 0L, response, smppServerTransactionImpl);
```

Comment:
what is this "0L" magic number? is the default timeout to be used while sending the actual req?if yes i think it shoudl be configurable from env...